## Importing pandas library as pd

In [39]:
import pandas as pd #importing pandas and assigning it to the alias pd

## Loading raw files into the dataframes

In [77]:
df_customers = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\olist_customers_dataset.csv") #loading customer dataset
df_geolocation = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\olist_geolocation_dataset.csv") #loading geolocation dataset
df_orders = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\olist_orders_dataset.csv") #loading orders dataset
df_products = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\olist_products_dataset.csv") #loading product dataset
df_sellers = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\olist_sellers_dataset.csv") #loading sellers dataset
df_order_payments = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\olist_order_payments_dataset.csv") #loading order payments dataset
df_order_reviews = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\olist_order_reviews_dataset.csv") #loading order reviews dataset
df_order_items = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\olist_order_items_dataset.csv") #loading order items dataset
df_product_category_translation = pd.read_csv(r"D:\Aswini\EDA\ecart\Data\raw\product_category_name_translation.csv") #loading product category name translation dataset

### Inspecting the loaded data using shape(for number of rows and columns), column(for column names as index), head(to display first few rows to understand the data structure) and info(for datatypes)

In [ ]:
# Checking shape columns and datatypes of Customer dataset 

print(f"=== Customer DF Shape ===\n{df_customers.shape}\n")#to identify the number of rows columns

print(f"=== Customer DF Columns ===\n{df_customers.columns}\n")  #To get the column names

print(f"=== Customer DF 1st 5 Rows ===\n{df_customers.head()} \n") #First 5 rows will be displayed

print(df_customers.info()) #The datatypes of each column will be obtained


=== Customer DF Shape ===
(99441, 5)

=== Customer DF Columns ===
Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='str')

=== Customer DF 1st 5 Rows ===
                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes       

In [42]:
# Checking shape columns and datatypes of Orders dataset 

print(f"=== Orders DF Shape ===\n{df_orders.shape}\n")#to identify the number of rows columns

print(f"=== Orders DF Columns ===\n{df_orders.columns}\n")  #To get the column names

print(f"=== Orders DF 1st 5 Rows ===\n{df_orders.head()} \n") #First 5 rows will be displayed

print(df_orders.info()) #The datatypes of each column will be obtained

=== Orders DF Shape ===
(99441, 8)

=== Orders DF Columns ===
Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

=== Orders DF 1st 5 Rows ===
                           order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-

In [43]:
# Checking shape columns and datatypes of Products dataset 

print(f"=== Products DF Shape ===\n{df_products.shape}\n")#to identify the number of rows columns

print(f"=== Products DF Columns ===\n{df_products.columns}\n")  #To get the column names

print(f"=== Products DF 1st 5 Rows ===\n{df_products.head()} \n") #First 5 rows will be displayed

print(df_products.info()) #The datatypes of each column will be obtained

=== Products DF Shape ===
(32951, 9)

=== Products DF Columns ===
Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='str')

=== Products DF 1st 5 Rows ===
                         product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0

In [44]:
# Checking shape columns and datatypes of Sellers dataset 

print(f"=== Sellers DF Shape ===\n{df_sellers.shape}\n")#to identify the number of rows columns

print(f"=== Sellers DF Columns ===\n{df_sellers.columns}\n")  #To get the column names

print(f"=== Sellers DF 1st 5 Rows ===\n{df_sellers.head()} \n") #First 5 rows will be displayed

print(df_sellers.info()) #The datatypes of each column will be obtained

=== Sellers DF Shape ===
(3095, 4)

=== Sellers DF Columns ===
Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='str')

=== Sellers DF 1st 5 Rows ===
                          seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP   

<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id       

In [45]:
# Checking shape columns and datatypes of Order payments dataset 

print(f"=== Order payments DF Shape ===\n{df_order_payments.shape}\n")#to identify the number of rows columns

print(f"=== Order payments DF Columns ===\n{df_order_payments.columns}\n")  #To get the column names

print(f"=== Order payments DF 1st 5 Rows ===\n{df_order_payments.head()} \n") #First 5 rows will be displayed

print(df_order_payments.info()) #The datatypes of each column will be obtained

=== Order payments DF Shape ===
(103886, 5)

=== Order payments DF Columns ===
Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='str')

=== Order payments DF 1st 5 Rows ===
                           order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45   

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data colum

In [46]:
# Checking shape columns and datatypes of Order Reviews dataset 

print(f"=== Order Reviews DF Shape ===\n{df_order_reviews.shape}\n")#to identify the number of rows columns

print(f"=== Order Reviews DF Columns ===\n{df_order_reviews.columns}\n")  #To get the column names

print(f"=== Order Reviews DF 1st 5 Rows ===\n{df_order_reviews.head()} \n") #First 5 rows will be displayed

print(df_order_reviews.info()) #The datatypes of each column will be obtained

=== Order Reviews DF Shape ===
(99224, 7)

=== Order Reviews DF Columns ===
Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='str')

=== Order Reviews DF 1st 5 Rows ===
                          review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                        

In [47]:
# Checking shape columns and datatypes of Order Items dataset 

print(f"=== Order Items DF Shape ===\n{df_order_items.shape}\n")#to identify the number of rows columns

print(f"=== Order Items DF Columns ===\n{df_order_items.columns}\n")  #To get the column names

print(f"=== Order Items DF 1st 5 Rows ===\n{df_order_items.head()} \n") #First 5 rows will be displayed

print(df_order_items.info()) #The datatypes of each column will be obtained

=== Order Items DF Shape ===
(112650, 7)

=== Order Items DF Columns ===
Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

=== Order Items DF 1st 5 Rows ===
                           order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004

In [48]:
# Checking shape columns and datatypes of Product category name translation dataset 

print(f"=== Product category name translation DF Shape ===\n{df_product_category_translation.shape}\n")#to identify the number of rows columns

print(f"=== Product category name translation DF Columns ===\n{df_product_category_translation.columns}\n")  #To get the column names

print(f"=== Product category name translation DF 1st 5 Rows ===\n{df_product_category_translation.head()} \n") #First 5 rows will be displayed

print(df_product_category_translation.info()) #The datatypes of each column will be obtained

=== Product category name translation DF Shape ===
(71, 2)

=== Product category name translation DF Columns ===
Index(['product_category_name', 'product_category_name_english'], dtype='str')

=== Product category name translation DF 1st 5 Rows ===
    product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor 

<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 3.5 KB
None


In [49]:
# Checking shape columns and datatypes of Geolocation dataset 

print(f"=== Geolocation DF Shape ===\n{df_geolocation.shape}\n")#to identify the number of rows columns

print(f"=== Geolocation DF Columns ===\n{df_geolocation.columns}\n")  #To get the column names

print(f"=== Geolocation DF 1st 5 Rows ===\n{df_geolocation.head()} \n") #First 5 rows will be displayed

print(df_geolocation.info()) #The datatypes of each column will be obtained

=== Geolocation DF Shape ===
(1000163, 5)

=== Geolocation DF Columns ===
Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='str')

=== Geolocation DF 1st 5 Rows ===
   geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP   

<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 

#### QUALITY ANALYSIS

#### Identifying the missing values in each dataframes using isnull().sum(), duplicated rows using duplicated().sum() and wrong datatype formats using dtypes

In [50]:
#Analysing the customer dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Customer data - Missing values ===\n{df_customers.isnull().sum()}\n")
print(f" === Customer data - Duplicated Rows ===\n{df_customers.duplicated().sum()}\n")
print(f" === Customer data - Datatypes of columns ===\n{df_customers.dtypes}")

 === Customer data - Missing values ===
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

 === Customer data - Duplicated Rows ===
0

 === Customer data - Datatypes of columns ===
customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object


## Customer Data Quality Report

### 🔹 Missing Values
- No missing values ✅  

### 🔹 Duplicates
- No duplicated rows ✅  

### 🔹 Incorrect Data Types
- No incorrect datatypes found ✅  


In [51]:
#Analysing the Orders dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Orders data - Missing values ===\n{df_orders.isnull().sum()}\n")
print(f" === Orders data - Duplicated Rows ===\n{df_orders.duplicated().sum()}\n")
print(f" === Orders data - Datatypes of columns ===\n{df_orders.dtypes}")

 === Orders data - Missing values ===
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

 === Orders data - Duplicated Rows ===
0

 === Orders data - Datatypes of columns ===
order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object


## Orders Data Quality Report

### Missing Values
| Column                        | Missing Count |
|-------------------------------|---------------|
| order_approved_at             | 160           |
| order_delivered_carrier_date  | 1,783         |
| order_delivered_customer_date | 2,965         |

### Duplicates
- No duplicated rows ✅  

### Incorrect Data Types
| Column                        | Current Type |
|-------------------------------|--------------|
| order_purchase_timestamp      | str          |
| order_approved_at             | str          |
| order_delivered_carrier_date  | str          |
| order_delivered_customer_date | str          |
| order_estimated_delivery_date | str          |


In [52]:
#Analysing the Products dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Products data - Missing values ===\n{df_products.isnull().sum()}\n")
print(f" === Products data - Duplicated Rows ===\n{df_products.duplicated().sum()}\n")
print(f" === Products data - Datatypes of columns ===\n{df_products.dtypes}")

 === Products data - Missing values ===
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

 === Products data - Duplicated Rows ===
0

 === Products data - Datatypes of columns ===
product_id                        str
product_category_name             str
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object


## Products Data Quality Report

### 🔹 Missing Values
| Column                     | Missing Count |
|----------------------------|---------------|
| product_category_name      | 610           |
| product_name_lenght        | 610           |
| product_description_lenght | 610           |
| product_photos_qty         | 610           |

### 🔹 Duplicates
- No duplicated rows ✅  

### 🔹 Incorrect Data Types
- None found ✅  


In [53]:
#Analysing the Sellers dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Sellers data - Missing values ===\n{df_sellers.isnull().sum()}\n")
print(f" === Sellers data - Duplicated Rows ===\n{df_sellers.duplicated().sum()}\n")
print(f" === Sellers data - Datatypes of columns ===\n{df_sellers.dtypes}")

 === Sellers data - Missing values ===
seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

 === Sellers data - Duplicated Rows ===
0

 === Sellers data - Datatypes of columns ===
seller_id                   str
seller_zip_code_prefix    int64
seller_city                 str
seller_state                str
dtype: object


## Sellers Data Quality Report

### 🔹 Missing Values
- No missing values ✅  

### 🔹 Duplicates
- No duplicated rows ✅  

### 🔹 Incorrect Data Types
- None found ✅  


In [54]:
#Analysing the Order payments dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Order payments data - Missing values ===\n{df_order_payments.isnull().sum()}\n")
print(f" === Order payments data - Duplicated Rows ===\n{df_order_payments.duplicated().sum()}\n")
print(f" === Order payments data - Datatypes of columns ===\n{df_order_payments.dtypes}")

 === Order payments data - Missing values ===
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64



 === Order payments data - Duplicated Rows ===
0

 === Order payments data - Datatypes of columns ===
order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object


## Order Payments Data Quality Report

### 🔹 Missing Values
- No missing values ✅  

### 🔹 Duplicates
- No duplicated rows ✅  

### 🔹 Incorrect Data Types
- None found ✅  


In [55]:
#Analysing the Order reviews dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Order reviews data - Missing values ===\n{df_order_reviews.isnull().sum()}\n")
print(f" === Order reviews data - Duplicated Rows ===\n{df_order_reviews.duplicated().sum()}\n")
print(f" === Order reviews data - Datatypes of columns ===\n{df_order_reviews.dtypes}")

 === Order reviews data - Missing values ===
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

 === Order reviews data - Duplicated Rows ===
0

 === Order reviews data - Datatypes of columns ===
review_id                    str
order_id                     str
review_score               int64
review_comment_title         str
review_comment_message       str
review_creation_date         str
review_answer_timestamp      str
dtype: object


## Order Reviews Data Quality Report

### Missing Values
| Column                  | Missing Count |
|-------------------------|---------------|
| review_comment_title    | 87,656        |
| review_comment_message  | 58,247        |

### Duplicates
- No duplicated rows ✅  

### Incorrect Data Types
| Column                   | Current Type |
|--------------------------|--------------|
| review_creation_date     | str          |
| review_answer_timestamp  | str          |

In [56]:
#Analysing the Order items dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Order items data - Missing values ===\n{df_order_items.isnull().sum()}\n")
print(f" === Order items data - Duplicated Rows ===\n{df_order_items.duplicated().sum()}\n")
print(f" === Order items data - Datatypes of columns ===\n{df_order_items.dtypes}")

 === Order items data - Missing values ===
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

 === Order items data - Duplicated Rows ===
0

 === Order items data - Datatypes of columns ===
order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object


## Order Items Data Quality Report

### Missing Values
- No missing values ✅  

### Duplicates
- No duplicated rows ✅  

### Incorrect Data Types
| Column              | Current Type |
|---------------------|--------------|
| shipping_limit_date | str          |

In [57]:
#Analysing the Product category name translation dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Product category name translation data - Missing values ===\n{df_product_category_translation.isnull().sum()}\n")
print(f" === Product category name translation data - Duplicated Rows ===\n{df_product_category_translation.duplicated().sum()}\n")
print(f" === Product category name translation - Datatypes of columns ===\n{df_product_category_translation.dtypes}")

 === Product category name translation data - Missing values ===
product_category_name            0
product_category_name_english    0
dtype: int64

 === Product category name translation data - Duplicated Rows ===
0

 === Product category name translation - Datatypes of columns ===
product_category_name            str
product_category_name_english    str
dtype: object


## Product Category Name Translation Data Quality Report

### Missing Values
- No missing values ✅  

### Duplicates
- No duplicated rows ✅  

### Incorrect Data Types
- None found ✅  


In [58]:
#Analysing the Geolocation dataset for missing values, duplicated rows and incorrect column datatypes
print(f" === Geolocation data - Missing values ===\n{df_geolocation.isnull().sum()}\n")
print(f" === Geolocation data - Duplicated Rows ===\n{df_geolocation.duplicated().sum()}\n")
print(f" === Geolocation data - Datatypes of columns ===\n{df_geolocation.dtypes}")

 === Geolocation data - Missing values ===
geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

 === Geolocation data - Duplicated Rows ===
261831

 === Geolocation data - Datatypes of columns ===
geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                   str
geolocation_state                  str
dtype: object


## Geolocation Data Quality Report

### Missing Values
- No missing values ✅  

### Duplicates
- 261,831 duplicated rows ⚠️  

### Incorrect Data Types
- None found ✅  


## DATA CLEANING

In [59]:
#Cleaning customers dataframe
def clean_customers(df: pd.DataFrame) -> pd.DataFrame:
    """
    Cleans the customers DataFrame:
    - Standardizes ZIP prefix to 5 digits (string)
    - Normalizes city (lowercase, stripped)
    - Normalizes state (uppercase, stripped)
    - Validates state length = 2
    - Removes helper column after check
    - Drops duplicate customer_id rows
    """
    df_clean = df.copy()

    # ZIP prefix as 5-digit string
    df_clean["customer_zip_code_prefix"] = (
        df_clean["customer_zip_code_prefix"].astype(str).str.zfill(5)
    )

    # Normalize city and state
    df_clean["customer_city"] = df_clean["customer_city"].str.lower().str.strip()
    df_clean["customer_state"] = df_clean["customer_state"].str.upper().str.strip()

    # Validate state length
    df_clean["state_length"] = df_clean["customer_state"].str.len()
    print(f"{df_clean["state_length"].value_counts()}\n")
    df_clean.drop(columns=["state_length"], inplace=True)

    # Drop duplicates by customer_id
    df_clean = df_clean.drop_duplicates(subset=["customer_id"])

    return df_clean

df_customers_clean = clean_customers(df_customers)

print(f"{df_customers_clean.dtypes}\n")
print(f"Unique customers: {df_customers_clean["customer_id"].nunique()}\n")
print(f"Shape: {df_customers_clean.shape}\n")
print(df_customers_clean.head())

state_length
2    99441
Name: count, dtype: int64

customer_id                 str
customer_unique_id          str
customer_zip_code_prefix    str
customer_city               str
customer_state              str
dtype: object

Unique customers: 99441

Shape: (99441, 5)

                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix          customer_city customer_state  
0                    14409                 franca             SP  
1                    09790  sao bernardo do campo             SP  
2                    01151              sao paulo             SP  
3                    08

In [60]:
df_customers_clean.to_csv(r"D:\Aswini\EDA\ecart\Data\cleaned\customers_clean.csv", index=False)

## Customers Data Cleaning Report

### Cleaning Summary
- **Zip Code Prefix** → normalized to 5 digits and converted to string  
- **Customer City** → standardized to lowercase and stripped of extra spaces  
- **Customer State** → standardized to uppercase and stripped of extra spaces  

### Missing Values
- No missing values ✅  

### Duplicates
- No duplicated rows ✅  

### Incorrect Data Types
- None found ✅  


In [61]:
#Analysing seller dataframe
from unidecode import unidecode
import re

def clean_sellers(df: pd.DataFrame) -> pd.DataFrame:
    """
    Cleans the sellers DataFrame:
    - Normalizes seller_city (lowercase, accents removed, stripped, cleaned of noise)
    - Replaces purely numeric city values with NaN
    - Normalizes seller_state (uppercase, stripped)
    - Normalizes seller_zip_code_prefix to 5-digit strings
    - Fixes specific known issue: 'lages  sc' → 'lages'
    - Drops duplicate seller_id rows
    """
    df_clean = df.copy()

    # Normalize seller_city
    df_clean["seller_city"] = (
        df_clean["seller_city"]
        .apply(lambda x: unidecode(str(x)).lower().strip())
        .str.replace(r"\s+", " ", regex=True)      # collapse spaces
        .str.replace(r"\(.*?\)", "", regex=True)   # remove parentheses content
        .str.replace(r"/.*", "", regex=True)       # remove anything after slash
        .str.replace(r"@.*", "", regex=True)       # remove email domains
        .str.replace(r"[^a-z\s]", "", regex=True)  # keep only letters + spaces
        .str.strip()
    )

    # Replace purely numeric city values with NaN
    df_clean.loc[df_clean["seller_city"].str.isnumeric(), "seller_city"] = None

    # Normalize seller_state
    df_clean["seller_state"] = (
        df_clean["seller_state"]
        .apply(lambda x: unidecode(str(x)).upper().strip())
    )

    # Normalize seller_zip_code_prefix to 5-digit strings
    df_clean["seller_zip_code_prefix"] = (
        df_clean["seller_zip_code_prefix"].astype(str).str.strip().str.zfill(5)
    )

    # Fix specific case
    df_clean["seller_city"] = df_clean["seller_city"].replace({"lages  sc": "lages"})

    # Drop duplicate sellers by seller_id
    df_clean = df_clean.drop_duplicates(subset=["seller_id"])

    return df_clean

df_sellers_clean = clean_sellers(df_sellers)

print(f"{df_sellers_clean.dtypes}\n")
print(f"Unique customers: {df_sellers_clean["seller_id"].nunique()}\n")
print(f"Shape: {df_sellers_clean.shape}\n")
print(df_sellers_clean.head())


seller_id                 str
seller_zip_code_prefix    str
seller_city               str
seller_state              str
dtype: object

Unique customers: 3095

Shape: (3095, 4)

                          seller_id seller_zip_code_prefix        seller_city  \
0  3442f8959a84dea7ee197c632cb2df15                  13023           campinas   
1  d1b65fc7debc3361ea86b5f14c68d2e2                  13844         mogi guacu   
2  ce3ad9de960102d0677a81f5d0bb7b2d                  20031     rio de janeiro   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                  04195          sao paulo   
4  51a04a8a6bdcb23deccc82b0b80742cf                  12914  braganca paulista   

  seller_state  
0           SP  
1           SP  
2           RJ  
3           SP  
4           SP  


In [62]:
df_sellers_clean.to_csv(r"D:\Aswini\EDA\ecart\Data\cleaned\sellers_clean.csv", index=False)

## Sellers Data Cleaning Report

### Cleaning Summary
- **Zip Code Prefix** → normalized to 5 digits and converted to string  
- **Seller City** → standardized to canonical lowercase city names, accents removed, extra spaces stripped, and noisy characters cleaned  
- **Seller State** → standardized to uppercase and stripped of extra spaces  
- **State Validation** → one numerical value found in state column, replaced with NaN  

### Missing Values
- No missing values ✅  

### Duplicates
- No duplicated rows ✅  

### Incorrect Data Types
- None found ✅  


In [63]:
from unidecode import unidecode

# -------------------------------
# Step 1: Clean Geolocation Table
# -------------------------------
def clean_geolocation(df: pd.DataFrame) -> pd.DataFrame:
    df_clean = df.copy()

    # Normalize ZIP prefix to 5-digit strings
    df_clean["geolocation_zip_code_prefix"] = (
        df_clean["geolocation_zip_code_prefix"].astype(str).str.zfill(5)
    )

    # Normalize state
    df_clean["geolocation_state"] = df_clean["geolocation_state"].str.upper().str.strip()

    # Normalize city
    df_clean["geolocation_city"] = (
        df_clean["geolocation_city"]
        .apply(lambda x: unidecode(str(x)).lower().strip().replace(".", ""))
    )

    # Drop duplicates
    df_clean = df_clean.drop_duplicates()

    return df_clean


# -------------------------------
# Step 2: Aggregate by ZIP Prefix
# -------------------------------
def aggregate_geolocation(df: pd.DataFrame) -> pd.DataFrame:
    return (
        df.groupby("geolocation_zip_code_prefix", as_index=False)
        .agg(
            lat=("geolocation_lat", "median"),
            lng=("geolocation_lng", "median"),
            city=("geolocation_city", lambda x: x.mode().iloc[0] if not x.mode().empty else None),
            state=("geolocation_state", lambda x: x.mode().iloc[0] if not x.mode().empty else None),
        )
    )


# -------------------------------
# Step 3: Build Summary for Missing ZIPs
# -------------------------------
def build_summary(df_customers, df_geo, window=10):
    # Normalize ZIPs
    df_customers["customer_zip_code_prefix"] = df_customers["customer_zip_code_prefix"].astype(str).str.zfill(5)
    df_geo["geolocation_zip_code_prefix"] = df_geo["geolocation_zip_code_prefix"].astype(str).str.zfill(5)

    # Find missing ZIPs
    missing_mask = ~df_customers["customer_zip_code_prefix"].isin(df_geo["geolocation_zip_code_prefix"])
    missing_zips = df_customers.loc[missing_mask, "customer_zip_code_prefix"].unique()

    summary_rows = []
    for zip_code in missing_zips:
        # Get city/state from customers
        cust_info = df_customers[df_customers["customer_zip_code_prefix"] == zip_code]
        city = cust_info["customer_city"].mode().iloc[0]
        state = cust_info["customer_state"].mode().iloc[0]

        # Get lat/lon from neighbors
        zip_int = int(zip_code)
        neighbors = df_geo[df_geo["geolocation_zip_code_prefix"].astype(int).between(zip_int - window, zip_int + window)]
        if not neighbors.empty:
            lat = neighbors["lat"].median()
            lng = neighbors["lng"].median()
        else:
            lat, lng = None, None

        summary_rows.append({
            "geolocation_zip_code_prefix": zip_code,
            "lat": lat,
            "lng": lng,
            "city": city,
            "state": state
        })

    return pd.DataFrame(summary_rows)


# -------------------------------
# Step 4: Build Summary for Sellers Missing ZIPs
# -------------------------------
def build_summary_sellers(df_sellers, df_geo, missing_zips, window=10):
    summary_rows = []
    for zip_code in missing_zips:
        # Get city/state from sellers
        seller_info = df_sellers[df_sellers["seller_zip_code_prefix"] == zip_code]
        city = seller_info["seller_city"].mode().iloc[0]
        state = seller_info["seller_state"].mode().iloc[0]

        # Get lat/lon from neighbors
        zip_int = int(zip_code)
        neighbors = df_geo[df_geo["geolocation_zip_code_prefix"].astype(int).between(zip_int - window, zip_int + window)]
        if not neighbors.empty:
            lat = neighbors["lat"].median()
            lng = neighbors["lng"].median()
        else:
            lat, lng = None, None

        summary_rows.append({
            "geolocation_zip_code_prefix": zip_code,
            "lat": lat,
            "lng": lng,
            "city": city,
            "state": state
        })

    return pd.DataFrame(summary_rows)


# -------------------------------
# Step 5: Full Workflow
# -------------------------------
# Clean geolocation
df_geolocations_clean = clean_geolocation(df_geolocation)

# Aggregate by ZIP
df_geolocations_clean_zip = aggregate_geolocation(df_geolocations_clean)

# Align with customers
customer_zips = set(df_customers_clean["customer_zip_code_prefix"].unique())
geo_zips = set(df_geolocations_clean_zip["geolocation_zip_code_prefix"].unique())

missing_zips_customers = customer_zips - geo_zips

print("Missing ZIPs (customers):", sorted(missing_zips_customers))

# Patch missing customer ZIPs
summary_df_customers = build_summary(df_customers_clean, df_geolocations_clean_zip)
df_geolocations_clean_zip_updated = pd.concat([df_geolocations_clean_zip, summary_df_customers], ignore_index=True)

# Align with sellers
df_sellers_clean["seller_zip_code_prefix"] = df_sellers_clean["seller_zip_code_prefix"].astype(str).str.zfill(5)
missing_zips_sellers = set(df_sellers_clean["seller_zip_code_prefix"].unique()) - set(df_geolocations_clean_zip_updated["geolocation_zip_code_prefix"].unique())

print("Missing ZIPs (sellers):", sorted(missing_zips_sellers))

# Patch missing seller ZIPs
summary_df_sellers = build_summary_sellers(df_sellers_clean, df_geolocations_clean_zip_updated, missing_zips_sellers)
df_geolocations_clean_zip_updated = pd.concat([df_geolocations_clean_zip_updated, summary_df_sellers], ignore_index=True)

# Final result
print("Final geolocation shape:", df_geolocations_clean_zip_updated.shape)
print(df_geolocations_clean_zip_updated.head(10))


Missing ZIPs (customers): ['02140', '06930', '07412', '07430', '07729', '07784', '08342', '08980', '11547', '12332', '12770', '13307', '17390', '19740', '25840', '25919', '27980', '28120', '28160', '28388', '28530', '28575', '28617', '28655', '29196', '29718', '29949', '35104', '35242', '35408', '36248', '36596', '36857', '36956', '37005', '38627', '38710', '39103', '41098', '41347', '42716', '42843', '43870', '44135', '48504', '49870', '55027', '55863', '56327', '56485', '57254', '58286', '58734', '59299', '59547', '61906', '62625', '62898', '64047', '64095', '64605', '65137', '65830', '67105', '68511', '68629', '70316', '70324', '70333', '70686', '70701', '70702', '70716', '71208', '71261', '71539', '71551', '71574', '71590', '71591', '71593', '71676', '71698', '71810', '71884', '71905', '71919', '71953', '71971', '71975', '71976', '71993', '71995', '71996', '72002', '72005', '72017', '72023', '72237', '72238', '72242', '72243', '72268', '72280', '72300', '72338', '72341', '72427', '

In [64]:
df_geolocations_clean_zip_updated.to_csv(r"D:\Aswini\EDA\ecart\Data\cleaned\geolocations_clean_zip_sql.csv", index=False)

## Geolocation Data Cleaning Report

### Cleaning Summary
- **Zip Code Prefix** → normalized to 5 digits and converted to string  
- **Geolocation State** → standardized to uppercase and stripped of extra spaces  
- **Geolocation City** → standardized to lowercase, accents removed, periods removed, and extra spaces stripped  
- **Duplicates** → 261,831 duplicated rows identified and removed  
- **Aggregation** → consolidated by ZIP prefix using median latitude/longitude and mode city/state  
- **ZIP Alignment** → missing ZIPs from Customers and Sellers inferred using nearest neighbors and appended to geolocation dataset  

### Missing Values
- No missing values ✅  

### Duplicates
- 261,831 duplicated rows ⚠️ (removed during cleaning)  

### Incorrect Data Types
- None found ✅  


In [65]:
#Product_category_name_translation dataframe data cleaning

def clean_product_category_translation(df_translation: pd.DataFrame, df_products: pd.DataFrame) -> pd.DataFrame:
    df_clean = df_translation.copy()

    # Normalize NaN to "unknown" in both datasets
    df_clean["product_category_name"] = df_clean["product_category_name"].fillna("unknown").astype(str)
    df_clean["product_category_name_english"] = df_clean["product_category_name_english"].fillna("unknown").astype(str)
    df_products["product_category_name"] = df_products["product_category_name"].fillna("unknown").astype(str)

    # Find missing categories (safe now, no float vs str issue)
    prod_cats = set(df_products["product_category_name"].unique())
    trans_cats = set(df_clean["product_category_name"].unique())

    missing_categories = prod_cats - trans_cats
    print("Categories missing in translation:", sorted(missing_categories))

    # Dynamically patch missing categories (default English = same as Portuguese)
    patch_rows = pd.DataFrame([
        {"product_category_name": cat, "product_category_name_english": cat}
        for cat in missing_categories
    ])

    df_clean = pd.concat([df_clean, patch_rows], ignore_index=True)
    print("Updated translation table shape:", df_clean.shape)

    return df_clean

df_product_category_translation_clean= clean_product_category_translation(
    df_product_category_translation,
    df_products
)

print(df_product_category_translation_clean.shape)

# Flag untranslated categories
df_untranslated = df_product_category_translation_clean[
        df_product_category_translation_clean["product_category_name"].str.strip().str.lower()
        == df_product_category_translation_clean["product_category_name_english"].str.strip().str.lower()
    ]
print("Untranslated categories:\n", df_untranslated[["product_category_name", "product_category_name_english"]])


Categories missing in translation: ['pc_gamer', 'portateis_cozinha_e_preparadores_de_alimentos', 'unknown']
Updated translation table shape: (74, 2)
(74, 2)
Untranslated categories:
                             product_category_name  \
19                                 consoles_games   
20                                          audio   
22                                     cool_stuff   
29                                       pet_shop   
31                                   market_place   
60                                     la_cuisine   
65                                   dvds_blu_ray   
71                                       pc_gamer   
72                                        unknown   
73  portateis_cozinha_e_preparadores_de_alimentos   

                    product_category_name_english  
19                                 consoles_games  
20                                          audio  
22                                     cool_stuff  
29                       

In [66]:
#Dictionary for untranslated categories
translation_dict = {
    "consoles_games": "Consoles & Games",
    "audio": "Audio",
    "cool_stuff": "Cool Stuff",
    "pet_shop": "Pet Shop",
    "market_place": "Marketplace",
    "la_cuisine": "Kitchenware",
    "dvds_blu_ray": "DVDs & Blu-ray",
    "portateis_cozinha_e_preparadores_de_alimentos": "Portable Kitchen Appliances",
    "unknown": "Unknown",
    "pc_gamer": "Gaming PC"
}

# Apply dictionary mapping to untranslated categories
# Update only those untranslated rows
df_product_category_translation_clean.loc[df_untranslated.index, "product_category_name_english"] = (
    df_untranslated["product_category_name"].map(lambda x: translation_dict.get(x, x))
)


print(df_product_category_translation_clean.loc[
    df_product_category_translation_clean["product_category_name"].isin(translation_dict.keys()),
    ["product_category_name", "product_category_name_english"]
])
print(df_product_category_translation_clean.shape)
print(df_product_category_translation_clean.tail(10))

                            product_category_name  \
19                                 consoles_games   
20                                          audio   
22                                     cool_stuff   
29                                       pet_shop   
31                                   market_place   
60                                     la_cuisine   
65                                   dvds_blu_ray   
71                                       pc_gamer   
72                                        unknown   
73  portateis_cozinha_e_preparadores_de_alimentos   

   product_category_name_english  
19              Consoles & Games  
20                         Audio  
22                    Cool Stuff  
29                      Pet Shop  
31                   Marketplace  
60                   Kitchenware  
65                DVDs & Blu-ray  
71                     Gaming PC  
72                       Unknown  
73   Portable Kitchen Appliances  
(74, 2)
                       

In [67]:
df_product_category_translation_clean.to_csv(r"D:\Aswini\EDA\ecart\Data\cleaned\product_category_translation_clean_sql.csv", index=False)

## Product Category Translation Cleaning Summary

### Steps Performed
- Normalized `NaN` values to `"unknown"` in both products and translation tables.
- Compared product categories with translation categories to detect missing entries.
- Dynamically patched missing categories by defaulting English = Portuguese.
- Flagged untranslated categories where Portuguese and English names were identical.
- Applied a dictionary mapping to fix untranslated categories.

### Dictionary Applied
- consoles_games → Consoles & Games  
- audio → Audio  
- cool_stuff → Cool Stuff  
- pet_shop → Pet Shop  
- market_place → Marketplace  
- la_cuisine → Kitchenware  
- dvds_blu_ray → DVDs & Blu-ray  
- portateis_cozinha_e_preparadores_de_alimentos → Portable Kitchen Appliances  
- unknown → Unknown  
- pc_gamer → Gaming PC  

### Outcome
- Missing categories patched dynamically.  
- Untranslated categories corrected using dictionary.  
- Already translated rows preserved.  
- Final translation table updated and ready for analysis.  

### Missing Values
- No missing values ✅  

### Duplicates
- No duplicated rows ✅  

### Incorrect Data Types
- None found ✅  

In [68]:
#Analysing order payments dataframe
def clean_order_payments(df_order_payments: pd.DataFrame) -> pd.DataFrame:
    """
    Cleans the Order Payments dataset and returns a single DataFrame
    ready for SQL loading:
    - Normalizes payment_type
    - Converts installments to numeric
    - Removes invalid values and installments
    - Caps extreme payment values at 99th percentile
    - Preserves payment_sequential and all original columns
    """

    df_clean = df_order_payments.copy()

    # 1. Normalize payment_type
    df_clean["payment_type"] = (
        df_clean["payment_type"]
        .str.lower().str.strip()
        .replace({
            "credit_card": "credit card",
            "boleto": "bank slip",
            "voucher": "voucher",
            "debit_card": "debit card"
        })
    )

    # 2. Convert installments to numeric
    df_clean["payment_installments"] = pd.to_numeric(
        df_clean["payment_installments"], errors="coerce"
    )

    # 3. Remove invalid payment values (<=0)
    df_clean = df_clean[df_clean["payment_value"] > 0]

    # 4. Cap extreme payment values at 99th percentile
    cap_value = df_clean["payment_value"].quantile(0.99)
    df_clean["payment_value"] = df_clean["payment_value"].clip(upper=cap_value)

    # 5. Remove invalid installments (<=0)
    df_clean = df_clean[df_clean["payment_installments"] > 0]

    # ✅ Keep payment_sequential and all other columns intact
    return df_clean


df_order_payments_clean_sql = clean_order_payments(df_order_payments)
print(df_order_payments_clean_sql.shape)
print(df_order_payments_clean_sql.dtypes)
print(df_order_payments.dtypes)


(103875, 5)
order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object
order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object


In [69]:
df_order_payments_clean_sql.to_csv(r"D:\Aswini\EDA\ecart\Data\cleaned\order_payments_clean.csv", index=False)

# Order Payments Data Cleaning Report

## 1. Cleaning Actions
- **Standardized Payment Types**
  - `credit_card` → credit card  
  - `boleto` → bank slip  
  - `voucher` → voucher  
  - `debit_card` → debit card  

- **Converted Data Types**
  - `payment_installments` converted to numeric (invalid entries coerced to NaN).  

- **Removed Invalid Records**
  - Dropped rows with `payment_value <= 0`.  
  - Dropped rows with `payment_installments <= 0`.  

- **Handled Outliers**
  - Capped `payment_value` at the 99th percentile to reduce extreme values.  

- **Duplicates Check**
  - Verified duplicates on (`order_id`, `payment_sequential`).  
 


In [70]:

def clean_order_items(df_order_items: pd.DataFrame) -> pd.DataFrame:
    """
    Cleans the Order Items dataset and returns a single DataFrame ready for SQL loading:
    - Converts shipping_limit_date to datetime
    - Ensures price and freight_value are numeric
    - Removes invalid prices and freight values
    - Caps extreme values at 99th percentile
    - Preserves all original columns (including order_item_id, order_id, etc.)
    """

    df_clean = df_order_items.copy()

    # 1. Convert shipping_limit_date to datetime
    df_clean["shipping_limit_date"] = pd.to_datetime(
        df_clean["shipping_limit_date"], errors="coerce"
    )

    # 2. Ensure price and freight_value are numeric
    df_clean["price"] = pd.to_numeric(df_clean["price"], errors="coerce")
    df_clean["freight_value"] = pd.to_numeric(df_clean["freight_value"], errors="coerce")

    # 3. Remove invalid prices (<=0)
    df_clean = df_clean[df_clean["price"] > 0]

    # 4. Cap extreme prices at 99th percentile
    cap_price = df_clean["price"].quantile(0.99)
    df_clean["price"] = df_clean["price"].clip(upper=cap_price)

    # 5. Remove invalid freight values (negative)
    df_clean = df_clean[df_clean["freight_value"] >= 0]

    # 6. Cap extreme freight values at 99th percentile
    cap_freight = df_clean["freight_value"].quantile(0.99)
    df_clean["freight_value"] = df_clean["freight_value"].clip(upper=cap_freight)

    # ✅ Return single cleaned DataFrame ready for SQL
    return df_clean



df_order_items_clean = clean_order_items(df_order_items)

print(df_order_items_clean[["price","freight_value"]].describe())
print(df_order_items_clean.shape)
print(df_order_items_clean.dtypes)


               price  freight_value
count  112650.000000  112650.000000
mean      115.274616      19.609478
std       136.379796      12.855061
min         0.850000       0.000000
25%        39.900000      13.080000
50%        74.990000      16.260000
75%       134.900000      21.150000
max       890.000000      84.520000
(112650, 7)
order_id                          str
order_item_id                   int64
product_id                        str
seller_id                         str
shipping_limit_date    datetime64[us]
price                         float64
freight_value                 float64
dtype: object


In [71]:
df_order_items_clean.to_csv(r"D:\Aswini\EDA\ecart\Data\cleaned\order_items_clean_final_sql.csv", index=False)

# Order Items Data Cleaning Report

## 1. Cleaning Actions
- **Converted Columns**
  - `shipping_limit_date` → converted to datetime (invalid entries coerced to NaT).
  - `price` and `freight_value` → converted to numeric (invalid entries coerced to NaN).

- **Removed Invalid Records**
  - Dropped rows with `price <= 0`.
  - Dropped rows with `freight_value < 0`.

- **Handled Outliers**
  - Capped `price` at the 99th percentile.
  - Capped `freight_value` at the 99th percentile.

- **Duplicates Check**
  - Verified duplicates across all columns.
  - Checked duplicates on (`order_item_id`, `order_id`).


In [72]:
import chardet

def clean_order_reviews(df_order_reviews: pd.DataFrame, export_path: str = None) -> pd.DataFrame:
    """
    Cleans the Order Reviews dataset and returns a single DataFrame ready for SQL loading:
    - Converts date columns to datetime
    - Fills missing comment/title values
    - Removes duplicate review_id rows (keeps last by answer timestamp)
    - Keeps only ER diagram columns
    - Optionally exports to CSV with UTF-8 encoding
    """

    df_clean = df_order_reviews.copy()

    # 1. Convert date columns to datetime
    df_clean["review_creation_date"] = pd.to_datetime(
        df_clean["review_creation_date"], errors="coerce"
    )
    df_clean["review_answer_timestamp"] = pd.to_datetime(
        df_clean["review_answer_timestamp"], errors="coerce"
    )

    # 2. Fill missing comments/titles
    df_clean["review_comment_message"] = df_clean["review_comment_message"].fillna("No comment")
    df_clean["review_comment_title"] = df_clean["review_comment_title"].fillna("No comment")

    # 3. Remove duplicate review_id rows (keep last by answer timestamp)
    df_clean = (
        df_clean.sort_values("review_answer_timestamp")
        .drop_duplicates(subset=["review_id"], keep="last")
    )

    # 4. Keep only ER diagram columns
    df_final = df_clean[[
        "review_id",
        "order_id",
        "review_score",
        "review_comment_message",
        "review_creation_date"
    ]]

    # 5. Optional export with encoding
    if export_path:
        df_final.to_csv(export_path, index=False, encoding="utf-8", errors="ignore")
        # Verify encoding with chardet
        with open(export_path, "rb") as f:
            print("Encoding detected:", chardet.detect(f.read(100000)))

    return df_final



df_order_reviews_clean_1 = clean_order_reviews(df_order_reviews, export_path=r"D:\Aswini\EDA\ecart\Data\cleaned\order_reviews_clean_sql_1.csv")

print(df_order_reviews_clean_1.shape)
print(df_order_reviews_clean_1.dtypes)


Encoding detected: {'encoding': 'utf-8', 'confidence': 0.8131328, 'language': 'pt', 'mime_type': 'text/plain'}
(98410, 5)
review_id                            str
order_id                             str
review_score                       int64
review_comment_message               str
review_creation_date      datetime64[us]
dtype: object


# Order Reviews Data Cleaning Report

## 1. Cleaning Actions
- **Converted Columns**
  - `review_creation_date` → converted to datetime (invalid entries coerced to NaT).
  - `review_answer_timestamp` → converted to datetime (invalid entries coerced to NaT).

- **Handled Missing Values**
  - `review_comment_message` → filled with `"No comment"`.
  - `review_comment_title` → filled with `"No comment"`.

- **Duplicate Resolution**
  - Identified duplicate `review_id` entries.
  - Sorted by `review_answer_timestamp` and retained the **latest record** per `review_id`.

- **Schema Alignment**
  - Retained only ER diagram columns:
    - `review_id`
    - `order_id`
    - `review_score`
    - `review_comment_message`
    - `review_creation_date`

## 2. Results
- **Original Rows**: 99,224  
- **Final Rows**: 98,410  
- **Rows Removed**: 814 duplicate entries dropped.  
- **Unique Reviews**: Each `review_id` now appears only once.  
- **Encoding**: Exported to CSV with `UTF-8` encoding (errors ignored), verified with `chardet`.



In [73]:
import numpy as np

def clean_products(df_products: pd.DataFrame, df_order_items: pd.DataFrame) -> pd.DataFrame:
    """
    Cleans the Products dataset and returns a single DataFrame ready for SQL loading:
    - Handles missing category names and product metadata
    - Imputes dimensions with category medians
    - Converts numeric columns to Int64 (nullable integer)
    - Clips unrealistic ranges for length
    - Adjusts zero values for weight and photos
    - Preserves schema with required columns only
    """

    df_clean = df_products.copy()

    # 1. Check if missing products appear in order_items
    missing_products = df_clean[df_clean["product_category_name"].isnull()]["product_id"]
    used_in_orders = df_order_items["product_id"].isin(missing_products).sum()
    print("Missing products used in orders:", used_in_orders)

    # 2. Fill missing category with 'unknown'
    df_clean["product_category_name"] = df_clean["product_category_name"].fillna("unknown")

    # 3. Fill missing name/description/photos with 0
    df_clean["product_name_lenght"] = df_clean["product_name_lenght"].fillna(0)
    df_clean["product_description_lenght"] = df_clean["product_description_lenght"].fillna(0)
    df_clean["product_photos_qty"] = df_clean["product_photos_qty"].fillna(0)

    # 4. Fill missing dimensions with category median
    dims = ["product_weight_g","product_length_cm","product_height_cm","product_width_cm"]
    df_clean[dims] = (
        df_clean.groupby("product_category_name")[dims]
        .transform(lambda x: x.fillna(x.median()))
    )

    # 5. Convert float columns to int
    cols_to_int = [
        "product_name_lenght",
        "product_description_lenght",
        "product_photos_qty",
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
    df_clean[cols_to_int] = df_clean[cols_to_int].astype("int64")

    # 6. Replace zero weights with median
    df_clean["product_weight_g"] = df_clean["product_weight_g"].replace(0, np.nan)
    df_clean["product_weight_g"] = df_clean["product_weight_g"].fillna(df_clean["product_weight_g"].median())

    # 7. Clip product_length_cm to realistic range
    df_clean["product_length_cm"] = df_clean["product_length_cm"].clip(lower=5, upper=200)

    # 8. Replace zero photos with 1
    df_clean["product_photos_qty"] = df_clean["product_photos_qty"].replace(0, 1)

    # 9. Keep only required columns
    df_final = df_clean[[
        "product_id",
        "product_category_name",
        "product_weight_g",
        "product_length_cm",
        "product_photos_qty"
    ]].copy()

    # 10. Convert numeric columns safely to Int64 (nullable integer)
    for col in ["product_weight_g", "product_length_cm", "product_photos_qty"]:
        df_final[col] = pd.to_numeric(df_final[col], errors="coerce").astype("Int64")

    # 11. Ensure string columns
    df_final["product_id"] = df_final["product_id"].astype(str)
    df_final["product_category_name"] = df_final["product_category_name"].astype(str)

    # ✅ Return cleaned DataFrame ready for SQL
    return df_final


df_products_clean_sql = clean_products(df_products, df_order_items)

print(df_products_clean_sql.dtypes)
print(df_products_clean_sql.shape)


Missing products used in orders: 0
product_id                 str
product_category_name      str
product_weight_g         Int64
product_length_cm        Int64
product_photos_qty       Int64
dtype: object
(32951, 5)


In [74]:
df_products_clean_sql.to_csv(r"D:\Aswini\EDA\ecart\Data\cleaned\products_clean_final_sql.csv", index=False)

# Products Data Cleaning Report

## 1. Cleaning Actions
- **Category Handling**
  - Filled missing `product_category_name` with `"unknown"`.
  - Checked if missing products were referenced in `order_items`.

- **Metadata Imputation**
  - Filled missing values in:
    - `product_name_lenght`
    - `product_description_lenght`
    - `product_photos_qty`
  - Replaced with `0` initially.

- **Dimensions Imputation**
  - Filled missing values in:
    - `product_weight_g`
    - `product_length_cm`
    - `product_height_cm`
    - `product_width_cm`
  - Used **category median** for imputation.

- **Numeric Adjustments**
  - Converted float columns to integers (`Int64` for nullable support).
  - Replaced zero weights with median values.
  - Clipped `product_length_cm` between **5 and 200 cm**.
  - Replaced zero photo counts with `1`.

- **Schema Alignment**
  - Retained only required columns:
    - `product_id`
    - `product_category_name`
    - `product_weight_g`
    - `product_length_cm`
    - `product_photos_qty`
  - Ensured `product_id` and `product_category_name` are strings.

## 2. Results
- **Invalid Values Fixed**: Zero weights replaced, photo counts adjusted.
- **Outliers Controlled**: Product length clipped to realistic range.
- **Schema Preserved**: Final dataset aligned with ER diagram requirements.
- **SQL Ready**: Cleaned DataFrame (`df_products_clean_sql`) with proper types:
  - Numeric columns → `Int64` (nullable integer).
  - String columns → `str`.



In [75]:
def clean_orders(df_orders: pd.DataFrame) -> pd.DataFrame:
    """
    Cleans the Orders dataset and returns a single DataFrame ready for SQL loading:
    - Converts timestamp columns to datetime
    - Imputes missing approval, carrier, and customer delivery dates using median lags
    - Caps extreme delivery delays and estimated delivery ranges
    - Fixes impossible sequences (delivery before purchase, estimate before purchase)
    - Retains only required columns with proper datatypes
    """

    df_clean = df_orders.copy()

    # 1. Convert string columns to datetime
    datetime_cols = [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
    for col in datetime_cols:
        df_clean[col] = pd.to_datetime(df_clean[col], errors="coerce")

    # 2. Impute missing approvals (delivered orders only)
    median_lag = (df_clean["order_approved_at"] - df_clean["order_purchase_timestamp"]).median()
    mask = (df_clean["order_approved_at"].isna()) & (df_clean["order_status"] == "delivered")
    df_clean.loc[mask, "order_approved_at"] = df_clean.loc[mask, "order_purchase_timestamp"] + median_lag

    # 3. Impute missing carrier dates (delivered orders only)
    median_shipping_lag = (
        df_clean.loc[df_clean["order_status"] == "delivered", "order_delivered_carrier_date"]
        - df_clean.loc[df_clean["order_status"] == "delivered", "order_purchase_timestamp"]
    ).median()
    mask = (df_clean["order_delivered_carrier_date"].isna()) & (df_clean["order_status"] == "delivered")
    df_clean.loc[mask, "order_delivered_carrier_date"] = (
        df_clean.loc[mask, "order_purchase_timestamp"] + median_shipping_lag
    )

    # 4. Impute missing customer delivery dates (delivered orders only)
    median_delivery_lag = (
        df_clean.loc[df_clean["order_status"] == "delivered", "order_delivered_customer_date"]
        - df_clean.loc[df_clean["order_status"] == "delivered", "order_delivered_carrier_date"]
    ).median()
    mask = (df_clean["order_delivered_customer_date"].isna()) & (df_clean["order_status"] == "delivered")
    df_clean.loc[mask, "order_delivered_customer_date"] = (
        df_clean.loc[mask, "order_delivered_carrier_date"] + median_delivery_lag
    )

    # 5. Delivery days calculation
    df_clean["delivery_days"] = (
        df_clean["order_delivered_customer_date"] - df_clean["order_purchase_timestamp"]
    ).dt.days

    # 6. Cap extreme delays at 60 days
    df_clean.loc[df_clean["delivery_days"] > 60, "order_delivered_customer_date"] = (
        df_clean["order_purchase_timestamp"] + pd.Timedelta(days=60)
    )

    # 7. Fix impossible delivery (before purchase)
    df_clean.loc[
        df_clean["order_delivered_customer_date"] < df_clean["order_purchase_timestamp"],
        "order_delivered_customer_date"
    ] = df_clean["order_purchase_timestamp"]

    # 8. Cap estimated delivery (max 30 days after purchase)
    df_clean.loc[
        df_clean["order_estimated_delivery_date"] > df_clean["order_purchase_timestamp"] + pd.Timedelta(days=30),
        "order_estimated_delivery_date"
    ] = df_clean["order_purchase_timestamp"] + pd.Timedelta(days=30)

    # 9. Fix impossible estimate (before purchase → set to +7 days)
    df_clean.loc[
        df_clean["order_estimated_delivery_date"] < df_clean["order_purchase_timestamp"],
        "order_estimated_delivery_date"
    ] = df_clean["order_purchase_timestamp"] + pd.Timedelta(days=7)

    # 10. Keep only required columns
    df_final = df_clean[[
        "order_id",
        "customer_id",
        "order_status",
        "order_purchase_timestamp",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]].copy()

    # 11. Enforce datatypes
    df_final["order_id"] = df_final["order_id"].astype(str)
    df_final["customer_id"] = df_final["customer_id"].astype(str)
    df_final["order_status"] = df_final["order_status"].astype(str)

    for col in ["order_purchase_timestamp", "order_delivered_customer_date", "order_estimated_delivery_date"]:
        df_final[col] = pd.to_datetime(df_final[col], errors="coerce")

    return df_final

df_orders_clean_final_sql = clean_orders(df_orders)

print(df_orders_clean_final_sql.dtypes)
print(df_orders_clean_final_sql.shape)


order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object
(99441, 6)


In [76]:
df_orders_clean_final_sql.to_csv(r"D:\Aswini\EDA\ecart\Data\cleaned\orders_clean_final_sql.csv", index=False)

# Orders Data Cleaning Report

## 1. Cleaning Actions
- **Datetime Conversion**
  - Converted all timestamp fields to datetime:
    - `order_purchase_timestamp`
    - `order_approved_at`
    - `order_delivered_carrier_date`
    - `order_delivered_customer_date`
    - `order_estimated_delivery_date`

- **Missing Value Imputation**
  - `order_approved_at`: Filled missing values for delivered orders using **purchase timestamp + median approval lag**.
  - `order_delivered_carrier_date`: Filled missing values for delivered orders using **purchase timestamp + median shipping lag**.
  - `order_delivered_customer_date`: Filled missing values for delivered orders using **carrier date + median delivery lag**.

- **Delivery Logic Adjustments**
  - Calculated `delivery_days` as difference between purchase and customer delivery.
  - Capped extreme delivery delays at **60 days**.
  - Fixed impossible deliveries (customer delivery before purchase → set to purchase date).

- **Estimated Delivery Adjustments**
  - Capped estimated delivery at **30 days after purchase**.
  - Fixed impossible estimates (before purchase → set to purchase date + 7 days).

- **Schema Alignment**
  - Retained only required columns:
    - `order_id`
    - `customer_id`
    - `order_status`
    - `order_purchase_timestamp`
    - `order_delivered_customer_date`
    - `order_estimated_delivery_date`
  - Enforced datatypes:
    - IDs and status → string
    - Timestamps → datetime

## 2. Results
- **Missing Dates Imputed**: Approval, carrier, and customer delivery dates filled for delivered orders.
- **Outliers Controlled**: Delivery delays capped at 60 days; estimated delivery capped at 30 days.
- **Logical Consistency**: No impossible sequences (delivery before purchase, estimate before purchase).
- **Final Dataset**: Cleaned DataFrame (`df_orders_clean_final_sql`) ready for SQL loading.

